### DISCUSSION TITLE
Bhop Detection Changes

### DISCUSSION THREAD

**Discussion by J-Tanzanite** [J-Tanzanite/Little-Anti-Cheat#57] [Ideas]
Just a place where we can discuss how to go about changing Bhop detections to be better and still be fast.

**Comment by J-Tanzanite** [comment]
## Things to change / Requirements:
- Bhop bans can't be set to be permanent, and will be limited to tops 3-6 months, any suggestion as to what the limit should be?
- It should check how often a player can do perfect Bhops in a row, and take into account +jump ticks.
- - But be configurable, easy to understand and pre-defined with good settings.
- Warning in chat for Admins when it suspects a player is doing too many perfect Bhops.
- A "max bhops" ConVar, for those who wish more control.
- The new system still has to be fast at detecting cheaters, I have a general idea as to how to do this, but suggestions are welcome.

Something else of note, there is a Macro detector coming in version 1.7.0, which is disabled by default (Because many server owners don't mind Macros, and this should be Opt-In, instead of Opt-Out).
And the Bhop detector should take Macros into account...
Sadly, due to how cheats bypass SMAC's Bhop detector, we can't make this that easy to bypass...
Plus, I've seen cheaters who do 40+ +jump ticks on Bhop...
The reason I know they are cheating is because they also got Aimbot detections (Several), Anti-Aim (Pitch AA, Duckspeed), spinbot in demo recording, Aimlock detections and chat-clear detections... So it deffo wasn't just a macro xD

**Reply by azalty** [reply]
"Bhop bans can't be set to be permanent, and will be limited to tops 3-6 months, any suggestion as to what the limit should be?"
I personally think we should let people chose the time they want, but we should warn them more.

"Plus, I've seen cheaters who do 40+ +jump ticks on Bhop..."
most people don't do more than 6 +jump inputs in 64 ticks servers due to input anti spam method the game has (ex: you probably know that spamming wheel will only send 3-6 inputs, never more. Scrolling slowly and consistently in comparison will make you able to send more than 20 inputs for each jump)
People with macros will probably have a constant amount of jump inputs. Can be 1 per jump if well coded, 2 per jump with SMAC bypass or X amount to always have 50% chance to hit bhop on any surface

I would like to get access to the source code of these macro to see how it's made

**Reply by J-Tanzanite** [reply]
I would still argue for a limit on ban time for Bhop, but more warnings isn't a bad idea.

---

Regarding the 40+, I know it wasn't a macro, not just because they got detected by every single other detection type, but also because they got this detection for Bhop with those same number several times.

So a little context here, there used to be a ban delay in Lilac, back when it was a private Anti-Cheat.
The reason for this was simple, since the source code wasn't available, cheaters couldn't check what got them banned, but they could try to figure it out through trial and error... But with a randomized delay, figuring out what they did became a lot harder.
Plus, in several cases, cheaters who got banned for Bhop would have a little more time on the server to get logged for other Cheat types, like Aimbot.

However this cheater wasn't banned, because I had disabled banning, just to see how far they'd go (I know, I am cruel, pls no hate)...
And they got:
`was detected and banned for Bhop (Jumps Presses: 40 | Bhops: 10)`
At least 20 times.

So it was most definitely a cheat thing, but I have no idea why their Bhop cheat was coded that way...
I've seen this specific number on other cheaters as well, although rarely... So I'm guessing the cheat itself isn't that popular.

**Reply by J-Tanzanite** [reply]
Sorry for the dreadfully slow reply/follow-up and progress, you know how life is during this time.
But I have worked a little on this issue and made some progress.

I've changed my mind regarding Bhop ban length time, I think you're right, server owners should still be allowed to set them to be permanent. More warnings should be enough.
I'm thinking I should have the warning be in two languages, one in English and one in Russian (if possible), as a lot of Russians do use Lilac but don't speak English...

---

One of the things I've experimented with, is making it so you can configure how many bhops in a row count as a detection, and how many detections are needed for a ban. And a Bhop detection expires after 10 minutes.
However, this ended up causing some issues, mainly with there being more ConVars than usual...

See, one thing I've noticed is that a lot of people don't quite understand how to configure plugins optimally, and especially so in regards to Anti-Cheat stuff... Heck, this has been such an issue in TF2 for a long time, I kinda wanna simplify things as much as possible, so it becomes harder to screw up.

---

On a smaller note, Bhop detections expiring after 10 minutes is a little... interesting.
That is how Aimbot detections work, they too expire after a while, to account for laggy players or network oddities for players who are on servers for hours on end.
If they didn't ever expire, then they would slowly accumulate over time and hit the ban threshold, which would be dumb.

However, in some games, you don't respawn upon dying, like in CS:GO (at least some of the gamemodes). Which means that the detection may expire by the time you finally respawn, in which case, the only way to get banned is to Bhop constantly all over the place, which isn't ideal.
But in games were you do respawn often, like TF2, having the timer be too long could be problematic as well.

So finding a good length for the timer is a little hard, especially if you consider how long it would take on average to get a 5 perfect consecutive Bhop vs a 10 long one... Because that would heavily affect the timer and the effectiveness of the detection.

---

Another thing is what we talked about in regards to the jump counter.
Cheats will usually do a 1 tick perfect jump, and sometimes 2 just so they can bypass SMAC.
Normal people and scrips may press +jump for a third or more tick... Now, trying to add an option to filter out Bhops where the +jump was spammed has been interesting...

Firstly, it creates the problem that cheats could bypass the detection simply by inserting more +jump ticks, which is how they bypassed SMAC in the first place... And I do not want to open that door (And that's ignoring the fact that some cheats currently do spam +jump)...
Even if I make that configurable, a lot of server owners won't get it, and may misunderstand it...
Especially since I'm not sure what I would call it. "Confidence"? "JumpPercentage"?
Either way, people will misunderstand it to believe it's a "false positive percentage" or "confidence/certainty" measurement that someone is cheating, which isn't what it would be measuring.

So I've been unsure if having such an option is wise.
It would help filter out manual Bhops, but would cause more issues in the long run.

---

I am looking at a different system tho, which would be easier to work with, fewer ConVars, less messy and be better long term.
Still a little unsure about it, but I'll report back in some days or so.

**Reply by azalty** [reply]
I added more detections methods to my private bhop anti cheat, I would like to share the private source code with you so we can see together if it is accurate and useful.

Of course, please don't republish this code or share it elsewhere, the goal of my private AC is to be effective, and having the source code open means people can get around more easily.

It's nowhere near great, but I think it does pretty accurate detections. (https://games.ludosoft.net/sourcebans/index.php?p=banlist&advSearch=0&advType=admin - check for 1 month bans) _NOTE: method 6&7 just got implemented, so they shouldn't have banned anyone yet. Every unban is because the user admitted he cheated and we forgave him because it was a simple mouse macro_

Could you send me your discord, or something to talk? :)

**Reply by J-Tanzanite** [reply]
Sure, I wouldn't mind.
But it would have to wait until after I update Lilac's Method for a few simple reasons.

The first one being ownership, if I had implemented a similar method to yours and I had seen your code, then cheaters could speculate on what you did.
If however, I have zero inspiration from your code, then they can't make any assumptions about similarity.
Another thing about ownership is that, if my method turns out to be similar to yours and I have seen your method (Assuming my method just happened to be similar to yours), then not giving you credit would be ethically wrong... But giving you credit would cause the issue I mentioned above.

For that reason, I should only take a look after I have updated Lilac's Bhop method.
But yeah, I don't mind giving you my take on it.

My Discord is J_Tanzanite#4402

**Reply by azalty** [reply]
I really doubt cheaters will try that much, I'm not that worried. You don't have to credit me as well, and I'm pretty sure your detection methods will be better. I own a single french server, and if cheat makers create a bypass, I would be happy to update my detection methods. Fighting them is just so entertaining, and it's quite impressive the amount of people that get banned.

I just made my plugin to make it a bit more fair, and detect people as a warning for them not to cheat. If it makes their life harder, I achieved my goal. Allows me to get better knowledge as well.

**Reply by J-Tanzanite** [reply]
I've been fairly busy as of late, but I have worked on the Bhop detection part.
I have experimented a bit with a few different methods.

I did experiment with a system where the bhops divided by the jump counter would give a "no spam %", but accounting for double jumps.
And using that as a multiplier of how certain it was that this wasn't a player who jumped, but rather a script (macro) or cheat.
However the results weren't satisfactory.

I also tried to add an option where server owners could set a threshold for this % instead, so they could ignore certain bhops...
But this has two main issues.
1. This is adding another convar which server owners will find confusing, because they may confuse it for a "likelyhood of cheating %"... Which is really wrong.
2. This basically allows cheats to bypass it more easily, even if it can be disabled.

Lastly, I considered a scoring system, but it was overly complicated.

I also have to consider these things:
1. The code must be easy to understand so server owners can trust it.
2. The code must be expandable/flexible, not only in usage, but also on a technical basis.
3. Updates MUST NOT BREAK PREVIOUS SYSTEMS. This is a requirement, I cannot make an update that breaks things (on purpose).
4. The interface must not be confusing / too complex. This basically means having 30 convars for a single feature, that's overkill and will always lead to misconfiguration (Even if you correct for it, it's an awful experience).

And that puts restraints on what I can do with Lilac.
Luckily for you, you don't have these restrictions, so making drastic changes is easier.

With these restraints, I decided to only add a single new feature, but change the previous bhop system to be more like my other detection methods.
Basically, there is now a counter for how many Bhop detections have been done in the past 10 minutes, and bhop detections expire after 10 minutes.
You can set how many bhop detections are needed to trigger a ban, with a minimum of 5 (might change that to 3, due to how different games handle respawning... Let alone gamemodes).
And you can set how many perfect consecutive bhops are needed to trigger a single detection.

The feature I added was a "max perfect consecutive bhops before an instant ban", this value can only be high (more on that later).

To prevent people from getting those mixed up, they are named as such:
lilac_bhop_min - Minimum perfect consecutive bhops needed to count as a single detection.
lilac_bhop_max - Maximum perfect consecutive bhops to trigger an instant ban.
lilac_bhop_total - How many bhop detections before triggering a ban.

The simplified and advanced modes won't exist anymore, and there is only "lilac_bhop -1 / 0 / 1" now.
-1 being log only, 0 being disabled, and 1 being enabled (with bans).

lilac_bhop_min has a minimum of 5 and a limit of 20.
lilac_bhop_max is by default off, although I am considering having the default value be 20, as that is high enough. The min value is 15 and there is no maximum value.

I am considering making it so people can disable the minimum detection system, and solely rely on maximum perfect consecutive bhops, might be nice.

**Reply by azalty** [reply]
I'm quite skeptical about this. Will have to see how it goes.

About the macro detection, please check the method I use on my plugin. It's very simple and effective

**Reply by J-Tanzanite** [reply]
I think I forgot to mention something in my previous post, sorry about that.

lilac_bhop_min would have a default value of 10, meaning that in order to get a ban with the default values, you would have to do 10 perfect consecutive bhops in a row to get a single detection... And then you would have to get 10 detections within the span of 10 minutes to get banned.
Doing that as legit should be fairly hard... And if it's a concern, it could be bumped from 10 consecutive perfect bhops to 12 ot 15, or you could up how many detections it would take to get banned within 10 minutes... Or both.

Or if the lilac_bhop_max value is set, then you could either get those 10 detections within 10 minutes or get 15-20 perfect consecutive perfect bhops in a row and get banned from that alone.

However... This is... Too slow, really.
The thing about the old "Advanced" mode was that it was really quick at handling cheaters who had 1 or 2 tick perfect inputs... And it would be preferable to keep that...
But adding that would mean adding 1 new ConVar again, making the bhop system complicated for server owners...
I know this may sound rude, but please don't take it that way... But back when I used to cheat, I would CONSTANTLY see TF2 players and admins why didn't understand cheats try to sound smart, getting just about everything wrong.

Some people would say that silentaim was when you could see players through walls... That was funny, or when they would say wallhacks means you can shoot through them (you can't in TF2, I know you sorta can in CS:GO) or the best of all... High sensitivity = silentaim...

Obviously most server owners are ok and understand the basics of cheats, but there are still those who don't.
And I don't want Lilac's interface to be overly complicated, it should be easy to use and configure... This is afterall one of the restrictions I have when developing it...

However... Yeah my previous post doesn't even make me feel satisfied, it's not good enough...
Back to the drawing board I guess...

**Reply by azalty** [reply]
10 bhops in a row is already rare enough, waiting for 10 warnings in 10mins is just too much

You can lower the min amount to 6 bhops for a warning and it could still be accurate

Or, you can ban after 2-3 warnings

Seriously, you know it's unlikely to get 10 bhops in a row. Why do you want to wait that much time to ban?

**Reply by azalty** [reply]
Another thing you could do is set config presets for users that don't know that much (ex: Medium for non bhop based gamemodes, High for bhop based gm and Very high for bhop only servers) while also allowing a "custom" option that will allow owners to set their own value. This will have a big warning saying that if they don't know exactly what they are doing, they should not touch this

**Reply by J-Tanzanite** [reply]
> 10 bhops in a row is already rare enough, waiting for 10 warnings in 10mins is just too much

I agree, hence why I don't like it.

> Seriously, you know it's unlikely to get 10 bhops in a row. Why do you want to wait that much time to ban?

To remove all doubt... But it's too slow, and it would take ages before someone gets banned.

---

I did consider dropping the ConVars and instead use commands, but I don't like the idea of hiding settings...
But considering this is something most server owners shouldn't touch in the first place... Perhaps it's fitting.

lilac_bhop could be used for the preset-configurations.
- 0 = Disabled.
- 1 = Low (Default perhaps), min = 6, max = 15, total = 3.
- 2 = Medium, min = 6, max = 17, total = 5.
- 3 = High, min = 7, max = 20, total = 5.
- 4 = Custom (Unlocks the command lilac_bhop_set command).

The negative values would mean the same thing, but be log-only.

Little unsure about these values, but since they are predefined configs for server owners to use, updating those through updates wouldn't break the third rule:
> 3. Updates MUST NOT BREAK PREVIOUS SYSTEMS. This is a requirement, I cannot make an update that breaks things (on purpose).

They wouldn't be breaking anything, just making things safer and accurate... So it's a nice idea.
I'd like to hear your idea on what the default preset values should be.

I think the command should work like this:
`lilac_bhop_set <min | max | total> <value>`
and not like:
`lilac_bhop_set <min> <max> <total> <value> <value> <value>`

This is because if you or a plugin wanna change a single setting and not change the others, you now have to remember each of the other settings, making changing anything a mess...
So setting the settings one by one is preferable.
Example:
`lilac_bhop_set min 6`
`lilac_bhop_set max 20`
`lilac_bhop_set total 4`

---

This system sounds way more simple and intuitive for server owners, props 2 u c:

**Reply by J-Tanzanite** [reply]
Ok, the previous comment is... Incorrect with the current 1.7.0 version. So here's how the new 1.7.0 system works:
#### Disclaimer: My keyboard is broken atm, so... You might see a few keys being pressed twice... Especially comma, semicolon, underscores and dashes... I'll get a new keyboard someday... Getting good keyboards in Norway is a pain ):

---

The ConVar "lilac__bhop" was changed, the minimum is -6 and the maximum is 6.
Negative values means "Log detections only, don't ban" (This is true for most cheat detection ConVars now).

0 = Disabled.
1 & 2 = Reserved/Placeholders, you can't pick these options, more on this later.
3 = Custom.
4 = Low.
5 = Medium (Default).
6 = High.

Each mode refer to how "aggressive" the Bhop check is, the higher it is, the faster it detects and bans cheaters, but also means it has a higher false positive rate.
The high mode is similar to the 1.6.X version, although slightly safer. It isn't recommended,, but it's there if people still want it (I don't).

1 & 2 are placed there so already existing configs don't break or suddenly begin acting really differently from what server owners expect.
If "lilac_bhop" is set to 1 or 2, it will tell you that those modes have been removed, and it will then set lilac_bhop to Medium.

Custom and Low modes have the same preset, except if you pick Custom, you can change your detection thresholds.
The Custom setting unlocks the command "lilac_bhop_set".

---

Here's how you'd use the command:
lilac_bhop_set <type> <value>
List of types are as follows:
min, ticks, max and total.



"Min" is the minimum consecutive perfect Bhops before it triggers a single detection.
I'll refer to these as "min-detections".
A "min-detection" expires after 10 minutes (This value is not set in stone, may change).

---

"Ticks" can be used to prevent min-detections from occurring.
It's a bit hard to explain, but it's a buffer.
Lets say you have the min set to 7, and the ticks set to 5.
The minimum jump ticks requires to do 7 bhops is - well - 7. And if ticks is set to 5, then the maximum jumpticks is 7+5 = 12.
This means that if you do 7 or more bhops and your jumpticks were higher or equal to 12, then you won't get a "min-detection".
You could use this to filter out legit spacebar presses, a lot of scroll (not all) and macros. Cheats tend to do 1 tick perfect inputs, and assuming no SMAC bypass is enabled, it should get a min-detection with ease.

If you want to account for SMAC bypasses, then you simply add the min value as well, like:
lilac_bhop_set min 7
lilac_bhop_set ticks 12
12 = 7 + 5. But when it checks for Min-detections, it will add the min value, so you're effectively doubling the jump ticks required, plus 5 as a buffer. Meaning the total buffer is 7+7+5 = 19.
I know this may be a little confusing at first, but it makes sense to make it this way.
Especially if you ONLY wanna ban for 7 bhops and 7 jumpticks, and ignore bhops with 8 or more jumpticks.
Which would be done like so:
lilac_bhop_set min 7
lilac_bhop_set ticks 0

If you want to disable the buffer entirely, you can do:
lilac_bhop_set ticks -1
This means it will not check jumpticks and check if you did "min" perfect consecutive bhops before giving a single min-detection.

---

"Max" means the maximum perfect consecutive Bhops done before doing an instant ban.
This value NEEDS to be high.
So lets say it was 15.
That means that on the 15'th perfect consecutive bhop, it will instantly ban the player.
Setting this value to 0 disables it, and it will only check for min-detections.

---

"total" is the total min-detections before banning.
This is pretty straight forward.
I noticed I made a little mistake in the pre-set configs, but I'll fix that soon :)

---

The default ban length for Bhop will remain as 1 month, I don't think I will change that.
Unless I get a good reason to do so.

---

When it comes to the preset values, they are as follows:

Low/Custom:
-    min = 7.
-    ticks = 7+5. // SMAC Bypass.
-    max = 20.
-    total = 5.

Medium:
-    min = 7.
-    ticks = -1. // Ignored, might change this to SMAC bypass, so legits are very unlikely to trigger... Although I think the "total"+"min = 7" more than enough covers legits.
-    max = 15.
-    total = 4.

High:
-    min = 5.
-    ticks = -1.
-    max = 10.
-    total = 2/4 (Fix coming soon. Props to you if you can spot what I refer to ;) )

These values may be changed in the future, these are just tmp values for now.

**Reply by azalty** [reply]
This seems pretty good, except the "high" setting which is still low in my opinion, setting it's "max" to 12 should do the trick though (also total = 2/3 seems low to me)

sad to see injection type detection leave with the default presets

As I already stated, imo ticks=min*2+.. is enough to catch legit people in my opinion, though I still need to do some statistics to get real numbers. Total = 5 is still really high, and setting it to 3 should still be accurate while being a lot more efficient

There doesn't seem to be an easy way to catch injection type bhops, normal frequent bhops ("min" cvar), and high amount of bhops ("max" cvar)

Trying to catch injection type bhops instantly (total = 1) would disable the other "min" and "total" approach

As I said, if you want accurate detections, you need a lot of testing and clearly LilAC is not a bhop AC but a general AC, and I think these new methods and configs are WAY better because the original did WAY too much false positives on Jailbreak/MG/... servers

Medium seems balanced. You COULD get 1 false positive in a year or so, but it's not a problem since as we said, it's probability versus efficiency, and server owners / admins know that it'll be accurate if it's a fishy or new account VS a veteran that is generally good at bhopping

**Reply by J-Tanzanite** [reply]
Firstly, nice name :D

Anyway, lets see if I've understood you correctly:

---

Regarding the High mode, fair. Max = 12 should be good.
When it comes to the total... Yeah that is low, perhaps 4 is low enough.
If admins wanna be "better safe than sorry" and overly paranoid, they could use the Custom mode and go bananas, I probably shouldn't enable such behavior through the High mode.

New High mode should be:
- min = 5.
- ticks = -1.
- max = 12.
- total = 3/4.

---

> As I already stated, imo ticks=min*2+.. is enough to catch legit people in my opinion

I'm going to assume you're talking about the internal number, as the buffer already adds the "min" value as it's necessary.

Well yes&no&yes... This gets murky.
People when they use spacebar tend to have 3-4+ tick inputs, assuming the tickrate is 64+
High tickrates have less problems with this, because humans' input aren't tied to a tickrate, but time... Cheats on the other hand are tick-based (usually).
Tho it's not unusual that people also achieve 2-3 tick inputs with spacebar. And those people may get issues with the SMAC bypass compensation... But there isn't really a whole lot of leeway here, as SMAC bypass is sometimes hardcoded into cheats... Not all, ofc, but catching those are important as well...

Another thing is scrolling, which does perfect 1 tick inputs. And people who use scroll to bhop either do 2-3 scrolls when they approach the ground or spam it constantly.
The former can get issues with the SMAC bypass compensation in Lilac's "ticks". The latter far less so, as they mimic macros, which also just spam wildly and will thus have a high JumpTicks value.

> Total = 5 is still really high, and setting it to 3 should still be accurate while being a lot more efficient

Agreed.
Since we're talking about the Low mode (I assume at least so), perhaps compensating for SMAC bypassing is... Well dumb.
It would make more sense for the Low mode to only catch the blatant cheaters which do near tick perfect inputs without SMAC bypassing.
So I guess the new Low mode should be:
- min = 7.
- ticks = 5. // NOT SMAC-Bypass compensated. This will ignore legits with 12 (7+5) or more jump ticks.
- max = 20. // Rare, but does happen on occation.
- total = 3.

---

I guess the Medium mode will be left unchanged for now?

---

One thing I do have to mention is Perfect Auto-Strafing.
Lilac used to detect those back when it was a private AC for my servers, but before release, I stripped Lilac of it (For a lot of reasons I've mentioned before).

It would be helpful to deal with those who also have SMAC Bypass in their Bhop.
And I think I've said in the past that in version 1.8.0 or so, I'll add it back.

I think that will have to wait a little further.
Version 1.7.0 is a big re-write and re-organization update, plus a few new misc features.
And I think once 1.7.0 is out, I'll just maintain version 1.7.0, fix issues and bugs as people report them, and have version 1.8.0 be mostly about optimizations and polish.

We'll see, the future is always uncertain.
But for now, 1.8.0 probably won't contain much interesting stuff.

**Reply by azalty** [reply]
> Firstly, nice name :D

Ty :)

> > I'm going to assume you're talking about the internal number, as the buffer already adds the "min" value as it's necessary.
> 
> Well yes&no&yes... This gets murky.
> People when they use spacebar tend to have 3-4+ tick inputs, assuming the tickrate is 64+

If I'm right, mouse wheel presses only do 1 tick input, and most of the time for a legit/good player they vary between 2/4, what I said is linked to this, I'm pretty sure spacebar presses last longer, and I should add this method to my AC as well to prevent false positives (as you said)

And as we're talking about odds, it is possible for someone to do in average 2 to 3 jump inputs per bhop  with scrolling (as you said), yet again with the current settings it's CLEARLY not a problem. Triggering only one detection with this should still be pretty rare, and if total is 3, it's clearly enough.

> I guess the Medium mode will be left unchanged for now?

Seems fine for me

> One thing I do have to mention is Perfect Auto-Strafing.

Now, I don't know a thing about auto-strafing, and I'm not good at strafing either (not a KZ player at all :eyes:)
Would be good to get that as well, I could learn how to detect that :3
